types/core/hooks.d.ts

```diff
@@ -12,7 +12,7 @@ declare class Hooks {
    * @param fn   - The callback function which should be triggered when the hook event occurs
    * @returns An ID number of the hooked function which can be used to turn off the hook later
    */
-  static on(hook: string, fn: Hooks.General): number;
+  static on<K extends string & keyof Hooks.Callbacks>(hook: K, fn: Hooks.Callbacks[K]): ReturnType<Hooks.Callbacks[K]>;
```

Comment:
might be worth adding an overload for plain strings, so the user isn't limited to known Hook keys but still benefits from the typing

Reply:
Yeah that might be another way forward. However I was thinking that we could achieve extensibility with letting the user extend `Hooks.Callbacks` with their own hooks and passing their extension as a generic type parameter. That's also where the intersection with `string` comes from. If you don't do that, it would allow `number` indexed objects as well.

Also after being done yesterday I realized that the return types might be wrong.